test.js

```diff
@@ -77,4 +77,159 @@ o('components', function () {
   })))).equals('<div><span>huhu</span></div>')
 })
 
+o('`this` in component', function () {
+  var oninit = o.spy()
+  var myComponent = {
+    oninit: function (vnode) {
+      oninit()
+      o(this).equals(vnode.state)('vnode.state should be the context in `oninit`')
+      o(this.foo).equals(undefined)('this.foo should be undefined initially')
+      this.foo = 5
+      o(vnode.state.bar).equals(4)('component properties should be copied to the state')
+    },
+    view: function (vnode) {
+      o(this).equals(vnode.state)('vnode.state should be the context in the view')
+      return m('div', 'hello')
+    },
+    onremove: function (vnode) {
+      o(this).equals(vnode.state)('vnode.state should be the context in `onremove`')
+    },
+    bar: 4
+  }
+
+  o(render([m(myComponent), m(myComponent)])).equals('<div>hello</div><div>hello</div>')
+
+  o(oninit.callCount).equals(2)('the component should have been initialized twice')
+})
+
+o('lifecycle hooks as attributes on elements', function () {
+  var initialized, removed
+  render(m('p',{
+    oninit: function (vnode) {
+      initialized = true
+      o(this).equals(vnode.state)('vnode.state should be the context in `oninit`')
+    },
+    onremove: function (vnode) {
+      removed = true
+      o(this).equals(vnode.state)('vnode.state should be the context in `onremove`')
+    }
+  }))
+  o(initialized).equals(true)('attr.oninit should run')
+  o(removed).equals(true)('attr.onremove should run')
+})
+
+o('lifecycle hooks as attributes on components', function () {
+  var attr_initialized, attr_removed, tag_initialized, tag_removed
+  var myComponent = {
+    oninit: function () {
+      o(attr_initialized).equals(true)('`attr.oninit()` should run before `tag.oninit()`')
+      tag_initialized = true
+    },
+    view: function(){return m('p','p')},
+    onremove: function () {
+      o(attr_removed).equals(true)('`attr.onremove()` should run before `tag.onremove()`')
+      tag_removed = true
+    }
+  }
+  o(render(m(myComponent, {
+    oninit: function(vnode) {
+      o(this).equals(vnode.state)('vnode.state should be the context in `attr.oninit`')
+      attr_initialized = true
+    },
+    onremove: function(vnode) {
+      o(this).equals(vnode.state)('vnode.state should be the context in `attr.onremove`')
+      attr_removed = true
+    }
+  }))).equals('<p>p</p>')
+  o(tag_initialized).equals(true)('tag.oninit should be called')
+  o(tag_removed).equals(true)('tag.onremove should be called')
```

Comment:
They are checked on line 136 and 140. The last two checks are there to make sure the former assertions were run.